homeassistant/components/met/weather.py

```diff
@@ -62,101 +58,80 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     if config.get(CONF_LATITUDE) is None:
         config[CONF_TRACK_HOME] = True
 
-    async_add_entities([MetWeather(config, hass.config.units.is_metric)])
+    await async_setup(hass, config, async_add_entities)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Add a weather entity from a config_entry."""
-    async_add_entities([MetWeather(config_entry.data, hass.config.units.is_metric)])
+    await async_setup(hass, config_entry.data, async_add_entities)
+
+
+async def async_setup(hass, config, async_add_entities):
+    """Create DataUpdateCoordinator and add entities."""
+    if config.get(CONF_TRACK_HOME, False):
+        unique_id = "home"
+    else:
+        unique_id = f"{config[CONF_LATITUDE]}-{config[CONF_LONGITUDE]}"
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            return await hass.data[DOMAIN][unique_id].fetch_data()
+        except Exception as err:
+            raise UpdateFailed(f"Update failed: {err}")
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=unique_id,
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(minutes=randrange(55, 65)),
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    hass.data[DOMAIN][unique_id].init_data()
+    await hass.data[DOMAIN][unique_id].fetch_data()
```

Comment:
You initialize the data here and again below.

I'd recommend using the `async_refresh` method of the coordinator instead.  This way you also take advantage of its debouncing behavior and logging.  If you need the data IO here to be successful to setup your entity, you will have to raise `PlatformNotReady` after this.  I'm not sure if this is the case.

Reply:
Having thought about it some more, I think this approach is fine too, but do you need to fetch twice during start up?